File: individual.py
"""
Individual class representing a solution in the genetic algorithm.
Optimized using NumPy for efficient array operations.
"""

import numpy as np
from typing import Optional
from fitness_calc import FitnessCalc


class Individual:
    """Represents a single solution (chromosome) in the genetic algorithm."""

    VALID = 1
    INVALID = 0
    DONT_KNOW = -1

    # Class-level taboo search variables
    _taboo: Optional[np.ndarray] = None
    _taboo_count: int = 100
    _taboo_set: bool = False

    def __init__(self, copy_from: Optional['Individual'] = None):
        """
        Initialize an individual.

        Args:
            copy_from: Optional individual to copy from
        """
        if copy_from is None:
            self.genes = np.zeros(FitnessCalc.number_of_vertices, dtype=np.int32)
            self.used_capacities = np.zeros(FitnessCalc.number_of_machines, dtype=np.float64)
            self.fitness = 0.0
            self.graph_cut_cost = 0.0
            self.validity = self.INVALID
            self.fitness_calculated = False
            self.best_gene_index = -1
        else:
            # Copy constructor
            self.genes = copy_from.genes.copy()
            self.used_capacities = copy_from.used_capacities.copy()
            self.fitness = copy_from.fitness
            self.graph_cut_cost = copy_from.graph_cut_cost
            self.validity = copy_from.validity
            self.fitness_calculated = copy_from.fitness_calculated
            self.best_gene_index = copy_from.best_gene_index

    def is_valid(self) -> int:
        """
        Check if the individual satisfies capacity constraints.

        Returns:
            VALID, INVALID, or DONT_KNOW
        """
        if self.validity != self.DONT_KNOW:
            return self.validity

        # Check if any machine is over capacity
        if np.any(self.used_capacities > FitnessCalc.M):
            self.validity = self.INVALID
        else:
            self.validity = self.VALID

        return self.validity

    def generate_valid_individual(self) -> None:
        """Generate a valid random individual satisfying all constraints."""
        free = FitnessCalc.M.copy()
        machines = np.arange(FitnessCalc.number_of_machines)

        for i in range(len(self.genes)):
            # Shuffle remaining machines
            available = FitnessCalc.number_of_machines
            for j in range(FitnessCalc.number_of_machines):
                value = np.random.randint(available - j)
                gene = machines[value]

                if free[gene] >= FitnessCalc.W[i]:
                    self.genes[i] = gene
                    free[gene] -= FitnessCalc.W[i]
                    self.used_capacities[gene] += FitnessCalc.W[i]
                    break

                # Swap to exclude this machine
                machines[value], machines[available - j - 1] = machines[available - j - 1], machines[value]

        self.validity = self.VALID

    def one_point_crossover(self, parent1: 'Individual', parent2: 'Individual') -> None:
        """
        Perform one-point crossover between two parents.

        Args:
            parent1: First parent
            parent2: Second parent
        """
        index = np.random.randint(parent1.size())
        bgene1 = parent1.best_gene_index
        bgene2 = parent2.best_gene_index

        self.used_capacities.fill(0)

        # First part from parent1
        for i in range(index):
            if i == bgene2:
                self.genes[i] = parent2.genes[bgene2]
            else:
                self.genes[i] = parent1.genes[i]
            self.used_capacities[self.genes[i]] += FitnessCalc.W[i]

        # Second part from parent2
        for i in range(index, parent1.size()):
            if i == bgene1:
                self.genes[i] = parent1.genes[bgene1]
            else:
                self.genes[i] = parent2.genes[i]
            self.used_capacities[self.genes[i]] += FitnessCalc.W[i]

        self.validity = self.DONT_KNOW

    def get_gene(self, index: int) -> int:
        """Get gene value at index."""
        return self.genes[index]

    def set_gene(self, index: int, value: int) -> None:
        """
        Set gene value at index and update fitness incrementally.

        Args:
            index: Gene index
            value: New machine assignment
        """
        prev_value = self.genes[index]
        if value == prev_value:
            return

        self.genes[index] = value
        self.validity = self.DONT_KNOW

        W = FitnessCalc.W
        M = FitnessCalc.M

        if not self.fitness_calculated:
            self.used_capacities[prev_value] -= W[index]
            self.used_capacities[value] += W[index]
            self.get_fitness()
        else:
            # Incremental fitness update
            if self.used_capacities[prev_value] > M[prev_value] and \
               self.used_capacities[prev_value] - W[index] <= M[prev_value]:
                self.fitness -= 1000000.0

            if self.used_capacities[value] <= M[value] and \
               self.used_capacities[value] + W[index] > M[value]:
                self.fitness += 1000000.0

            self.used_capacities[prev_value] -= W[index]
            self.used_capacities[value] += W[index]

            C = FitnessCalc.C
            B = FitnessCalc.B

            # Calculate fitness change
            # Add costs for new assignment
            mask_new = self.genes != value
            fitness_plus = np.sum(C[index, :] * B[self.genes, value] * mask_new)

            # Subtract costs for old assignment
            mask_old = self.genes != prev_value
            fitness_minus = np.sum(C[index, :] * B[self.genes, prev_value] * mask_old)

            self.fitness += fitness_plus - fitness_minus
            self.graph_cut_cost += fitness_plus - fitness_minus

    def size(self) -> int:
        """Return the number of genes."""
        return len(self.genes)

    def get_fitness(self) -> float:
        """
        Calculate and return fitness value.

        Returns:
            Fitness value (lower is better)
        """
        if self.fitness_calculated:
            return self.fitness

        # Penalty for capacity violations
        self.fitness = np.sum(np.maximum(0, self.used_capacities - FitnessCalc.M) > 0) * 1000000.0

        # Calculate graph cut cost using vectorized operations
        C = FitnessCalc.C
        B = FitnessCalc.B
        self.graph_cut_cost = 0.0

        for i in range(FitnessCalc.number_of_vertices):
            for j in range(i + 1, FitnessCalc.number_of_vertices):
                if self.genes[i] != self.genes[j]:
                    cost = C[i, j] * B[self.genes[i], self.genes[j]]
                    self.fitness += cost
                    self.graph_cut_cost += cost

        self.fitness_calculated = True
        return self.fitness

    def get_graph_cut_cost(self) -> float:
        """Get the graph cut cost (without penalties)."""
        if not self.fitness_calculated:
            self.get_fitness()
        return self.graph_cut_cost

    @classmethod
    def get_taboo_of_index(cls, index: int) -> int:
        """Get taboo value for a gene index."""
        if cls._taboo_set:
            return cls._taboo[index]
        else:
            cls._set_taboo()
            cls._taboo_set = True
            return 0

    @classmethod
    def set_taboo_of_index(cls, index: int) -> None:
        """Set taboo value for a gene index."""
        if not cls._taboo_set:
            cls._set_taboo()
            cls._taboo_set = True

        cls._taboo[index] = cls._taboo_count
        cls._taboo_count += 1

    @classmethod
    def _set_taboo(cls) -> None:
        """Initialize taboo array."""
        cls._taboo = np.zeros(FitnessCalc.number_of_vertices, dtype=np.int32)

    @classmethod
    def get_taboo_count(cls) -> int:
        """Get current taboo count."""
        return cls._taboo_count

    def __str__(self) -> str:
        """String representation of the individual."""
        lines = [f"(Vertex {i} ==> Machine {self.genes[i]})" for i in range(self.size())]
        return '\n'.join(lines)

File: java/Individual.java
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.Random;

public class Individual {

    private int[] genes;
    private double[] usedCapacities;       // used amount of each machine
    private double fitness;
    private double graphCutCost;
    private int validity;
    private boolean fitnessCalculated;

    private int bestGeneIndex;

    public static final int VALID = 1;
    public static final int INVALID = 0;
    public static final int DONTKNOW = -1;

    private static int [] taboo = new int[FitnessCalc.numberOfVertex];
    private static int tabooCount = 100;
    private static boolean tabooSet = false;

    public Individual() {
        genes = new int[FitnessCalc.numberOfVertex];
        usedCapacities = new double[FitnessCalc.numberOfMachines];

        validity = INVALID;
        fitnessCalculated = false;

        bestGeneIndex = -1;
    }

    public Individual(Individual you) {

        genes = new int[FitnessCalc.numberOfVertex];
        for (int i = 0; i < genes.length; i++) {
            genes[i] = you.getGene(i);
        }
        double[] youUsedCapacities = you.getUsedCapacities();

        usedCapacities = new double[FitnessCalc.numberOfMachines];
        for (int i = 0; i < FitnessCalc.numberOfMachines; i++) {
            usedCapacities[i] = youUsedCapacities[i];
        }
        this.fitness = you.getFitness();
        this.graphCutCost = you.getGraphCutCost();

        this.validity = you.isValid();
        fitnessCalculated = you.getFitnessCalculated();

        bestGeneIndex = you.getBestGeneIndex();
    }

    public int isValid() {

        if (validity != DONTKNOW) return validity;

        double[] M = FitnessCalc.M;

        for (int i = 0; i < FitnessCalc.numberOfMachines; i++) {
            if (usedCapacities[i] > M[i]) {
                validity = INVALID;
                return validity;
            }
        }
        validity = VALID;
        return validity;
    }

    public void generateValidIndividual() {

        double[] free = new double[FitnessCalc.numberOfMachines];
        int[] machine = new int[FitnessCalc.numberOfMachines];

        for (int i = 0; i < free.length; i++) {
            free[i] = FitnessCalc.M[i];
            machine[i] = i;
        }
        int value, temp, gene;
        Random random = new Random();
        for (int i = 0; i < size(); i++) {
            for (int j = 0; j < FitnessCalc.numberOfMachines; j++) {

                value = random.nextInt(FitnessCalc.numberOfMachines - j);
                gene = machine[value];
                if (free[gene] >= FitnessCalc.W[i]) {
                    genes[i] = gene;
                    free[gene] -= FitnessCalc.W[i];
                    usedCapacities[gene] += FitnessCalc.W[i];
                    break;
                }
                temp = machine[FitnessCalc.numberOfMachines - j - 1];
                machine[FitnessCalc.numberOfMachines - j - 1] = machine[value];
                machine[value] = temp;
            }
        }
        this.validity = VALID;
    }

    public void onePointCrossover(Individual indiv1, Individual indiv2) {
        Random random = new Random();
        int index = random.nextInt(indiv1.size());
        int bgene1 = indiv1.getBestGeneIndex(), bgene2 = indiv2.getBestGeneIndex();
        int i;

        for(int j=0; j<FitnessCalc.numberOfMachines; j++) usedCapacities[j] = 0;

        for (i = 0; i < index; i++) {
            if(i==bgene2){
                genes[i] =  indiv2.getGene(bgene2);
                usedCapacities[genes[i]] += FitnessCalc.W[i];
                continue;
            }
            int gene = indiv1.getGene(i);
            genes[i] = gene;
            usedCapacities[gene] += FitnessCalc.W[i];
        }
        for (; i < indiv1.size(); i++) {
            if(i==bgene1){
                genes[i] =  indiv1.getGene(bgene1);
                usedCapacities[genes[i]] += FitnessCalc.W[i];
                continue;
            }
            int gene = indiv2.getGene(i);
            genes[i] = gene;
            usedCapacities[gene] += FitnessCalc.W[i];
        }

        validity = DONTKNOW;
    }

    public int getGene(int index) {
        return genes[index];
    }

    public void setGene(int index, int value) {

        int prevValue = genes[index];
        if (value == prevValue) {
            return;
        }
        genes[index] = value;
        validity = DONTKNOW;

        double[] W = FitnessCalc.W;
        double[] M = FitnessCalc.M;

        if (!fitnessCalculated) {
            usedCapacities[prevValue] -= W[index];
            usedCapacities[value] += W[index];
            getFitness();
        } else {
            if (usedCapacities[prevValue] > M[prevValue] && usedCapacities[prevValue] - W[index] <= M[prevValue])
                fitness -= 1000000.0;

            if (usedCapacities[value] <= M[value] && usedCapacities[value] + W[index] > M[value])
                fitness += 1000000.0;

            usedCapacities[prevValue] -= W[index];
            usedCapacities[value] += W[index];

            double[][] C = FitnessCalc.C;
            double[][] B = FitnessCalc.B;

            double fitnessPlus = 0, fitnessMinus = 0;

            for (int i = 0; i < FitnessCalc.numberOfVertex; i++) {
                if (genes[i] != value) {
                    fitnessPlus += C[i][index] * B[genes[i]][value];
                }
                if (genes[i] != prevValue) {
                    fitnessMinus += C[i][index] * B[genes[i]][prevValue];
                }
            }

            fitness += fitnessPlus;
            fitness -= fitnessMinus;

            graphCutCost += fitnessPlus;
            graphCutCost -= fitnessMinus;

//            double ffitness = FitnessCalc.getFitness(this);
//            double graphCC = FitnessCalc.getGraphCutCost(this);
//
//            if (Math.abs(ffitness - fitness) > 0.5 || Math.abs(graphCC - graphCutCost) > 0.5) {
//
//
//                BigDecimal bd1 = new BigDecimal(ffitness);
//                BigDecimal bd2 = new BigDecimal(fitness);
//                BigDecimal bd3 = new BigDecimal(graphCC);
//                BigDecimal bd4 = new BigDecimal(graphCutCost);
//
//                System.out.println("---------------");
//                System.out.println("---------------");
//                System.out.println("---------------");
//                System.out.println("---------------");
//
//                System.out.println("ffitness: " + bd1);
//                System.out.println("fitness : " + bd2);
//                System.out.println("graphCC: " + bd3);
//                System.out.println("graphcc: " + bd4);
//
//                System.out.println("---------------");
//                System.out.println("---------------");
//                System.out.println("---------------");
//                System.out.println("---------------");
//
//                System.out.println();
//            }
        }
        validity = DONTKNOW;
    }

    public int size() {
        return genes.length;
    }

    public double getFitness() {

        if (!fitnessCalculated) {
            fitness = 0;
            for (int i = 0; i < FitnessCalc.numberOfMachines; i++) {
                if (usedCapacities[i] > FitnessCalc.M[i]) {
                    fitness += 1000000.0;
                }
            }
            double[][] C = FitnessCalc.C;
            double[][] B = FitnessCalc.B;
            graphCutCost = 0;
            for (int i = 0; i < FitnessCalc.numberOfVertex; i++) {
                for (int j = i + 1; j < FitnessCalc.numberOfVertex; j++) {
                    if (genes[i] != genes[j]) {
                        double d = (C[i][j] * B[genes[i]][genes[j]]);
                        fitness += d;
                        graphCutCost += d;

                    }
                }
            }
            fitnessCalculated = true;
        }
        return fitness;
    }

    public double getGraphCutCost() {
        if (!fitnessCalculated) {
            getFitness();
        }
        return graphCutCost;
    }

    public boolean getFitnessCalculated() {
        return fitnessCalculated;
    }


    @Override
    public String toString() {
        StringBuffer strBuff = new StringBuffer();
        for (int i = 0; i < size(); i++) {
            strBuff.append("(Vertex " + i + " ==> Machine " + getGene(i) + ")\n");
        }
        return strBuff.toString();
    }

    public double[] getUsedCapacities() {
        return usedCapacities;
    }

    public static int getTabooOfIndex(int index){
        if(tabooSet) return taboo[index];
        else {
            setTaboo();
            tabooSet = true;
            return 0;
        }
    }

    public static void setTabooOfIndex(int index) {
        if (tabooSet) {
            taboo[index] = tabooCount;
        } else {
            setTaboo();
            tabooSet = true;
            taboo[index] = tabooCount;
        }
        tabooCount++;
    }


    private static void setTaboo(){
        for(int i=0; i<FitnessCalc.numberOfMachines; i++) taboo[i] = 0;
    }

    public static int getTabooCount() {
        return tabooCount;
    }

    public int getBestGeneIndex() {
        return bestGeneIndex;
    }

    public void setBestGeneIndex(int bestGene) {
        this.bestGeneIndex = bestGene;
    }
}